daq/stream_monitor.py

```diff
@@ -145,16 +149,18 @@ def event_loop(self):
                 LOGGER.error('Monitoring exception in callback: %s', e)
                 LOGGER.exception(e)
             self.log_monitors()
-            self.dirty = False
-            fds = self.poller.poll(self.timeout_ms)
+            fds = self.poller.poll(self.timeout_sec * 10e3 if self.timeout_sec else None)
             LOGGER.debug('Monitoring found fds %s', fds)
             if fds:
                 for fd, event in fds:
-                    self.process_poll_result(event, fd)
-                    if self.dirty:
-                        LOGGER.debug('Monitoring set modified, re-issuing poll...')
-                        break
-            else:
-                LOGGER.debug('Monitoring no fds found')
-                return True
+                    if fd in self.callbacks:
+                        self.process_poll_result(event, fd) #Monitoring set could be modified
+            # check for timeouts
+            frozen_callbacks = copy.copy(self.callbacks)
+            for fd in frozen_callbacks:
+                name, _, _, on_error, timeout = frozen_callbacks[fd]
+                if timeout and datetime.fromtimestamp(time.time()) >= timeout:
+                    self.error_handler(fd, TimeoutError("Timed out."), name, on_error)
+                    if fd in self.callbacks:
+                        self.forget(fd)
```

Comment:
This seems indeterminate here -- in that error_handler *should* or must remove the callback, otherwise it's a logical error. I would go the other way... assert that the fd is in fact removed from the callbacks and throw an error if not. -- unless there is some fundamental reason why that doesn't happen.